### PR TITLE
MolGan layers

### PR DESCRIPTION
**Pull request by MiloszGrabski**
# MolGAN layers

## Next step in implementation of simplified MolGAN generative model originally published by Nicola De Cao and Thomas Kipf ["MolGAN: An implicit generative model for small molecular graphs"](https://arxiv.org/abs/1805.11973)

Added all required layers and unit tests.
Rather than creating a new file I have added them to [models/layer.py](https://github.com/deepchem/deepchem/blob/master/deepchem/models/layers.py).
There are several flake8 errors within the original file.
I have removed some by removing whitespaces, but there are a few that remain.
My code passes flake8 and yapf without problems.

Have some issues with mypy:
deepchem\feat\molecule_featurizers\molgan_featurizer.py:143: error: Incompatible return value type (got "Optional[GraphMatrix]", expected "GraphMatrix")"
 --Was resolved by updating master branch version     

I would welcome suggestion on how to sort it out.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by MiloszGrabski** [comment]
Ok, I have managed to locate the issue but not sure how to sort it out.
This solves the issue: return graph  #if (degree > 0).all() else None
But I would still be able to return None if object is invalid for further filtering.
I could always set-up some sort of a flag in object, but would rather avoid doing this.

File: deepchem/models/layers.py [hunk 1]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
     return mol_features
 
 
+class GraphConvolutionLayer(tf.keras.layers.Layer):
```
Comment (rbharath) [inline deepchem/models/layers.py:373]: Would these layers be useful beyond `MolGan`? If not, perhaps better to rename `MolGanGraphConvolutionLayer`

File: deepchem/models/layers.py [hunk 2]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+    Invoke this layer
+
+    Parameters
+    ----------
+    inputs: list
+      List of two input matrices, adjacency tensor and node features tensors
+      in one-hot encoding format.
+    training: bool
+      Should this layer be run in training mode.
+      Typically decided by main model, influences things like dropout.
```
Comment (rbharath) [inline deepchem/models/layers.py:439]: Can you add a 

```
Returns
--------
...
```
section here for the return values?

File: deepchem/models/layers.py [hunk 3]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+  Graph convolution layer used in MolGAN model.
+  MolGAN is a WGAN type model for generation of small molecules.
+  Not used directly, higher level layers like MultiGraphConvolutionLayer use it.
+  This layer performs basic convolution on one-hot encoded matrices containing
+  atom and bond information. This layer also accepts three inputs for the case
+  when convolution is performed more than once and results of previous convolution
+  need to used. It was done in such a way to avoid creating another layer that
+  accepts three inputs rather than two. The last input layer is so-called
+  hidden_layer and it hold results of the convolution while first two are unchanged
+  input tensors.
```
Comment (rbharath) [inline deepchem/models/layers.py:384]: Could you add an
```
Example
---------
...
```
section here with a usage example?

File: deepchem/models/layers.py [hunk 4]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+    return config
+
+
+class GraphAggregationLayer(tf.keras.layers.Layer):
+  """
+  Graph Aggregation layer used in MolGAN model.
+  MolGAN is a WGAN type model for generation of small molecules.
+  Performs aggregation on tensor resulting from convolution layers.
+  Given its simple nature it might be removed in future and to 
+  GraphEncoderLayer.
```
Comment (rbharath) [inline deepchem/models/layers.py:484]: Can you add a 
```
Example
--------
...
```
 section with a usage example?

File: deepchem/models/layers.py [hunk 5]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+    Invoke this layer
+
+    Parameters
+    ----------
+    inputs: list
+      List of two input matrices, adjacency tensor and node features tensors
+      in one-hot encoding format.
+    training: bool
+      Should this layer be run in training mode.
+      Typically decided by main model, influences things like dropout.
```
Comment (rbharath) [inline deepchem/models/layers.py:623]: Can you add a 
```
Returns
-------
...
```
section here?

File: deepchem/models/layers.py [hunk 6]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+
+
+class MultiGraphConvolutionLayer(tf.keras.layers.Layer):
+  """
+  Multiple pass convolution layer used in MolGAN model.
+  MolGAN is a WGAN type model for generation of small molecules.
+  It takes outputs of previous convolution layer and uses
+  them as inputs for the next one.
+  It simplifies the overall framework, but might be moved to
+  GraphEncoderLayer in the future in order to reduce number of layers.
```
Comment (rbharath) [inline deepchem/models/layers.py:563]: Can you add an example section here?

File: deepchem/models/layers.py [hunk 7]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+
+    config = super(GraphAggregationLayer, self).get_config()
+    config["units"] = self.units
+    config["activation"] = self.activation
+    config["dropout_rate"] = self.dropout_rate
+    config["edges"] = self.edges
+    return config
+
+
+class MultiGraphConvolutionLayer(tf.keras.layers.Layer):
```
Comment (rbharath) [inline deepchem/models/layers.py:556]: Perhaps better to rename `MolGanMultiGraphConvolutionLayer`

File: deepchem/models/layers.py [hunk 8]
```diff
@@ -370,6 +370,380 @@ def call(self, inputs):
...
+
+    config = super(GraphConvolutionLayer, self).get_config()
+    config["activation"] = self.activation
+    config["dropout_rate"] = self.dropout_rate
+    config["units"] = self.units
+    config["edges"] = self.edges
+    return config
+
+
+class GraphAggregationLayer(tf.keras.layers.Layer):
```
Comment (rbharath) [inline deepchem/models/layers.py:478]: Perhaps better to rename `MolGanGraphAggregationLayer`

File: deepchem/models/tests/test_molgan_layers.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    model = keras.Model(
+        inputs=[adjacency_tensor, node_tensor], outputs=[hidden_tensor])
+
+    assert model.output_shape == (None, vertices, second_convolution_unit)
+    assert layer.units == units
+    assert layer.activation == activations.tanh
+    assert layer.edges == 5
+    assert layer.dropout_rate == 0.0
+
+  def test_graph_encoder_later(self):
```
Comment (rbharath) [inline deepchem/models/tests/test_molgan_layers.py:70]: typo: "later" -> "layer"

**Review by rbharath — COMMENTED** [review]
Looks pretty good! Have some minor comments with requests for more documentation and suggested name changes for the layers

File: deepchem/models/layers.py (hunk 1 above)
Comment (MiloszGrabski) [reply deepchem/models/layers.py:373]: It is possible as it is based on Relational-GCN [(Schlichtkrull et al., 2017)](https://arxiv.org/abs/1703.06103).
I will live it up to you to decide. Let me know and I will change names for layers.

Comment (rbharath) [reply deepchem/models/layers.py:373]: Hmm, how about `RelationalGraphConvLayer` and we add a citation to the Schlichtkrull et al. paper in the references?

File: deepchem/models/layers.py (hunk 7 above)
Comment (rbharath) [reply deepchem/models/layers.py:601]: Should this be `RelationMultiGraphConvolutionalLayer` then?

File: deepchem/models/layers.py [hunk 10]
```diff
@@ -370,6 +370,471 @@ def call(self, inputs):
...
+  """
+  Main learning layer used by MolGAN model.
+  MolGAN is a WGAN type model for generation of small molecules.
+  It role is to further simplify model.
+  This layer can be manually built by stacking graph convolution layers
+  followed by graph aggregation.
+
+  Example
+  --------
+  def create_discriminator(adjacency_tensor, node_tensor):
```
Comment (rbharath) [inline deepchem/models/layers.py:726]: To have this render correctly you'd want to do this as
```
Example
------
>>> def create_discriminator(...)
...
```
That is, with the `>>>` added before the code sample

**Review by rbharath — COMMENTED** [review]
Getting very close now! Have a suggested name change and one more minor comment about docs style

File: deepchem/models/layers.py (hunk 1 above)
Comment (MiloszGrabski) [reply deepchem/models/layers.py:373]: I have checked code I could find and it is a bit different. I think using MolGAN is least confusing. I will change the code to MolGAN prefix.

File: deepchem/models/layers.py (hunk 10 above)
Comment (MiloszGrabski) [reply deepchem/models/layers.py:726]: OK, I have added per your suggestion and indented the code to match.
Should be fine now. 
I have been thinking and I will probably modify layers slightly once everything is in place (layers, model etc.)
I have been thinking about improving overall flow of the code.
This is something I will pick your brain about once we release the code.

**Comment by MiloszGrabski** [comment]
It fails doctest on builds but doctest on my machine raises no errors, any idea?

**Comment by rbharath** [comment]
Oh, hmm, that's weird. It looks like `deepchem/feat/tests/test_materials_featurizers.py::TestMaterialFeaturizers::test_sine_coulomb_matrix` and `deepchem/data/tests/test_json_loader.py::test_json_loader` are failing which shouldn't have anything to do with this PR. (deepchem/models/tests/test_gat.py::test_gat_regression is failing too but I think that's stochastic).

@ncfrey Would you have any ideas here? Could this maybe be an upstream version bump in pymatgen or the like?

**Comment by ncfrey** [comment]
> Oh, hmm, that's weird. It looks like `deepchem/feat/tests/test_materials_featurizers.py::TestMaterialFeaturizers::test_sine_coulomb_matrix` and `deepchem/data/tests/test_json_loader.py::test_json_loader` are failing which shouldn't have anything to do with this PR. (deepchem/models/tests/test_gat.py::test_gat_regression is failing too but I think that's stochastic).
> 
> @ncfrey Would you have any ideas here? Could this maybe be an upstream version bump in pymatgen or the like?

This could definitely be a `pymatgen` version issue. They do releases with breaking changes pretty frequently. We can probably pin to the [v2021.1.28](https://github.com/materialsproject/pymatgen/releases/tag/v2021.1.28) release or earlier to fix this.

**Comment by MiloszGrabski** [comment]
So I suppose you can merge then?

**Comment by peastman** [comment]
We need to get the pymatgen errors fixed before we can merge.  When Github Actions encounters a failure, it stops running.  If any test fails, the later stages never get run at all.  So we can only merge on a completely clean CI report.

**Comment by rbharath** [comment]
Kicking off a new run with the `pymatgen` issues fixed. Let's see how we do :)

**Comment by MiloszGrabski** [comment]
Same error:
=========================== short test summary info ============================
FAILED deepchem/data/tests/test_json_loader.py::test_json_loader - assert (0,...
FAILED deepchem/feat/tests/test_materials_featurizers.py::TestMaterialFeaturizers::test_sine_coulomb_matrix

**Comment by rbharath** [comment]
Hmm, perhaps we need to rebase here? The fix PR was all green so these should be fixed in theory

**Comment by MiloszGrabski** [comment]
Given that my code did not touch base at all, it should work.
At least that is what I have concluded from reading, I personally never done rebasing.

**Comment by MiloszGrabski** [comment]
looks like more issues with the base

**Comment by rbharath** [comment]
Hmm, this is puzzling, most tests are passing but we're seeing a couple of failures on the docs build and windows install step I haven't seen before.

@peastman @ncfrey Would either of you have any ideas here?

Sorry for the trouble @MiloszGrabski. We're still adjusting to the Github actions CI build and need to get our CI in a clean state

**Comment by peastman** [comment]
Both of the Windows builds failed while trying to install netcdf4.  Perhaps an update broke it?  We could try just rerunning one of them to see if the problem has already been fixed.

**Comment by MiloszGrabski** [comment]
@rbharath No problem. Solving problems is part of development. There is no point in rushing if it will be broken.

**Comment by MiloszGrabski** [comment]
Let me know once you fix base and will try rebase again.

**Comment by ncfrey** [comment]
Hi @MiloszGrabski, issues with base should be solved now so if you rebase it should fix the errors.

**Comment by MiloszGrabski** [comment]
@ncfrey done, keeping fingers crossed

**Comment by MiloszGrabski** [comment]
@ncfrey @rbharath Documents still fail giving this error: 
Exception occurred:
  File "/usr/share/miniconda3/envs/deepchem/lib/python3.7/site-packages/sphinx/builders/html/__init__.py", line 1039, in <lambda>
    ctx['css_files'].sort(key=lambda js: js.priority)
AttributeError: 'str' object has no attribute 'priority'
The full traceback has been saved in /tmp/sphinx-err-1_libhy1.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:26: recipe for target 'html' failed
make: *** [html] Error 2

**Comment by rbharath** [comment]
I think we might be good to merge this PR in. The docs issue is something we need to fix but doesn't look to be related to this PR.

@ncfrey @peastman Let me know if you have any comments here. Will hold off on merging for a day or two in case you want to take another look

**Comment by rbharath** [comment]
Going to go ahead and merge this PR since no additional comments!

**Pull request merged**